Add /me endpoint that returns user's info

jessebrennan opened the pull request:
This adds the '/me' endpoint and returns the requested info. Is there any security risk in giving back the access token in json @coverbeck?

jessebrennan committed:
Add /me endpoint that returns user's info

coverbeck commented:
> Is there any security risk in giving back the access token in json @coverbeck?

I don't think so. There is a requirement to surface the token in the UI, DataBiosphere/cgp-boardwalk/65, so I think we have to return it.

Also, since we only support https, the body is not visible.

BTW, the access token already is in the cookie, as you know, and not encrypted, so this offers the same level of safety/danger as that.

coverbeck commented on app.py:
```diff
@@ -314,6 +315,62 @@ def check_session(cookie):
+
+    In addition, if the access token was refreshed, the new access token
+    should be sent back in the session cookie.
+    """
```
Solid comments! Makes is easy to follow.

coverbeck commented on app.py:
```diff
@@ -314,6 +315,62 @@ def check_session(cookie):
+    """
+    returns information about the user making the request.
+
+    If authentication is enabled for the deployment, and there is no
```
This should be `required for the deployment` -- even if authentication is not required, we will still allow users to optionally log in.

jessebrennan replied:
fixed

coverbeck commented on app.py:
```diff
@@ -314,6 +315,62 @@ def check_session(cookie):
+    access token, or it is expired and cannot be renewed, then return a
+    401.
+
+    If authentication is not enabled for the deployment, and there is no
```
required not enabled

jessebrennan replied:
fixed

jessebrennan commented:
> BTW, the access token already is in the cookie, as you know, and not encrypted, so this offers the same level of safety/danger as that.

Oh yeah. duh.

coverbeck commented on app.py:
```diff
@@ -314,6 +315,62 @@ def check_session(cookie):
+        new_google_access_token()
+        resp = get_user_info_from_token()
+    if resp.status_code != 200:
+        return 'Invalid access code could not be refreshed', 401
```
1. It's an invalid (expired) refresh token, isn't it? Should you say the refresh token has expired?
2. I'm wondering if you shouldn't clear the refresh token from this cookie at this point, although I suppose there's the possibility that the request failed, for example, because the network connection was down or some unexpected situation. That's why I'm wondering if you can tell how it failed.

jessebrennan replied:
Yeah, I wasn't completely sure what to do here. If the token is invalid we get a 401.  Other responses I'm not sure about. I could pass it on the error if it's not 401. 

Otherwise you suggest clearing the token? Then prompting them to log back in again? We should discuss further.

coverbeck replied:
Minor point, I don't think you should say `access code`, I think you should `access token`.  I also think this is worth logging.

If it's an invalid token and not refreshable, then I think you should  clear both tokens. Returning 401 is enough; because this will be called via AJAX, you can't redirect to login from here.

The exceptions to clearing could be if you got 5xx from Google, which indicates an error on their side.

coverbeck commented on app.py:
```diff
@@ -263,6 +263,7 @@ def new_google_access_token():
         'client_secret': Auth.CLIENT_SECRET,
     }
     resp = oauth.refresh_token(Auth.TOKEN_URI, refresh_token=refresh_token, **extra)
+    current_user.access_token = resp['access_token']
```
What happens if this call fails? Does it raise an exception, or do you need to check for it here first?

jessebrennan replied:
Sure I suppose if it failed it would raise an exception, but I don't see why it ever would.

coverbeck replied:
It could fail because:

1. The refresh token could have expired
2. The refresh token could have been revoked (this should actually be a testable case, as you need to revoke the refresh token on logout)
3. Vague system issues -- even services like Google have outages, network connections fail or get saturated, etc.

jessebrennan replied:
Oh I see, I thought you were referring to that specific line. Yeah I can document and make the usage more robust.

coverbeck replied:
Ah, I see -- yes, I meant you may need error handling between lines 265 and 266.

jessebrennan replied:
Actually, looking at the oauth.refresh_token code, if it fails to get a new token, ~it just returns the old token. So I'll just have to check for that case in the code that calls this function.~  an OAuth2Error is raised.

coverbeck reviewed (changes requested):
I realize some of the comment changes may be my fault, as I may have written up incorrectly, but reading it again, I think they should be updated.

Looks good in general though! Nice and small set of changes.

jessebrennan committed:
Update docstring

jessebrennan committed:
Improve error handling and messages

Now if token can't be refreshed this is explicitly logged and the
session is cleared.

If some other error happens a 401 is still returned, but message
reflects it better

coverbeck commented on app.py:
```diff
@@ -314,6 +320,68 @@ def check_session(cookie):
+    if resp.status_code != 200:
+        return 'Failed to get user info: ' + resp.text, 401
+    user_data = resp.json()
+    output = dict((k, user_data[k]) for k in ('name', 'email'))
```
I was looking at this one last time, and I realized I overlooked this in my first review -- I think you should check against the whitelist if there is one.

One could argue that you could only have gotten here if you were on the whitelist in the first place, and that's true, but if somebody logged in, then got taken off the whitelist, they would still have access. So I think you should check.

The check may overlap with the work you are doing for the auth endpoint, so if you want to tackle it there instead, that would be OK (although maybe add a comment here saying you need to authorize).

jessebrennan replied:
I was working on the auth endpoint on a sub-branch of this branch. I may just close this PR and reopen both changes together. The auth endpoint is requiring some refactoring of this stuff anyway.

coverbeck replied:
Sounds good

coverbeck commented on app.py:
```diff
@@ -314,6 +320,68 @@ def check_session(cookie):
+            return 'Could not refresh access token: ' + e.message, 401
+        resp = get_user_info_from_token()
+    if resp.status_code != 200:
+        return 'Failed to get user info: ' + resp.text, 401
```
I wonder if you should pop the access and refresh tokens here as well, like you do a few lines above?

I guess it depends what other status codes you can get on line 365... If the token is invalid, are you always going to get a 400?

jessebrennan replied:
I was thinking against popping it again in case It's a 500 error. I'm not really sure what's possible but I figured if something unforeseen goes wrong this will catch it, pass the error on. 

If the token is invalid you should get a 400 error.

coverbeck replied:
I agree, you shouldn't pop it if it's a 5xx error -- that indicates some problem on the Google side that might get fixed by them. So if those are the only two types of status codes this is OK. Maybe add a comment though, since it wasn't clear to me, and won't necessarily be to the next person who reads it.

jessebrennan replied:
It may be a 500 error, but idk if it's possible there will be an other type of error. I can still add a comment to clarify though. Should I pop for any breed of 4xx errors (other than 400)?
